### ISSUE TITLE
prediction fails after resuming bot

### ISSUE THREAD

**Issue by cristianmtr** [RasaHQ/rasa#5645]
**Rasa version**: Rasa 1.9.5

**Rasa SDK version** (if used & relevant): 1.9.0

**Rasa X version** (if used & relevant): n/a

**Python version**: 3.6

**Operating system** (windows, osx, ...): Linux Ubuntu

**Issue**:

1. bot is paused in order for customer support to talk with customer
2. bot is resumed via the execute action HTTP endpoint
3. intent is triggered on behalf of user through custom endpoint (to send recommendations to the user). This is always done **after** the resume action has completed
4. the intent doesn't trigger the intended action. 

OR

1. bot is paused in order for customer support to talk with customer
2. bot is resumed via the execute action HTTP endpoint
3. user sends message to bot (This is always done **after** the resume action has completed)
4. the intent doesn't trigger the intended action. 

**NOTEs**: 

1. The intent **does trigger** the intended action if we wait *X* amount of seconds before sending **3.**. This seems to be a **timing issue**.

When it works it does detect the trigger mapping:

```
2020-04-15 11:23:11 DEBUG    rasa.core.policies.mapping_policy  - The predicted intent 'receive_notify_new_apartments' is mapped to  action 'action_notify_new_apartments' in the domain.
```

2. We tried this with two intents that both have triggers in the domain, but are also included in stories. 
3. We deploy the bot on Elastic Beanstalk, use 2 instances, have one single PSQL tracker store, and one Redis lock store (if it's a lock race condition or similar issue).

**Error (including full traceback)**:
```
2020-04-15 11:22:04 DEBUG    rasa.core.policies.mapping_policy  - There is no mapped action for the predicted intent, 'receive_notify_new_apartments'
[...]
2020-04-15 11:22:04 DEBUG    rasa.core.policies.memoization  - There is no memorised next action
2020-04-15 11:22:04 DEBUG    rasa.core.policies.form_policy  - There is no active form
2020-04-15 11:22:04 DEBUG    rasa.core.policies.fallback  - NLU confidence threshold met, confidence of fallback action set to core threshold (0.3).
2020-04-15 11:22:04 DEBUG    rasa.core.policies.ensemble  - Predicted next action using policy_1_AlpacaKerasPolicy
2020-04-15 11:22:04 DEBUG    rasa.core.processor  - Predicted next action 'action_listen' with confidence 0.98.
```

**Command or request that led to error**:

User intent being correctly identified by the NLU, but then not being predicted correctly.

**Content of configuration file (config.yml)** (if relevant):

Not relevant. If required, I can provide them.

```yml

```

**Content of domain file (domain.yml)** (if relevant):

Not relevant. If required, I can provide them.

```yml

```

Adding @nbeuchat

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @wochinge will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by wochinge** [comment]
Hi @cristianmtr ,

thanks for the detailed description.

> bot is resumed via the execute action HTTP endpoint

I assume this triggers a custom action, which then adds a event to resume the bot?

> and one Redis lock store (if it's a lock race condition or similar issue).

Could you please run the rasa instances with `--debug`, go through the described flow and share the logs? That would be super helpful, thanks 🙏

**Comment by nbeuchat** [comment]
Hi @wochinge just sharing the code of the action we use to resume the bot. Please note that we use `UserUtteranceReverted` to remove all intents (it surprisingly works) said after the bot was paused just before reactivating it. We have to do this as the user inputs while the bot is paused are still inserted in the tracker and breaks everything upon unpausing. 

```
from rasa_sdk import Action
from rasa_sdk.events import *

from actions.ensure_state import ensure_state_all


class ActionResumeBot(Action):
    def name(self):
        return "action_resume_bot"

    @ensure_state_all
    def run(self, dispatcher, tracker, domain):
        if tracker.is_paused():
            dispatcher.utter_message(template="utter_resume_bot")

            uttered_while_paused = False
            for event in reversed(tracker.events_after_latest_restart()):
                if event.get("event") == "pause":
                    break
                elif event.get("event") == "user":
                    uttered_while_paused = True
                    break

            if uttered_while_paused:
                # Remove all utterances while the bot was paused and resume bot
                return [UserUtteranceReverted(), ConversationResumed()]
            else:
                # Remove action_pause_bot and resume bot
                return [ConversationResumed()]
        else:
            return [ActionReverted()]
``` 

@cristianmtr can share more of the logs

**Comment by cristianmtr** [comment]
@wochinge I will email the logs to support@rasa.com and reference the issue number.

**Comment by wochinge** [comment]
Thanks for the logs @cristianmtr 🙏 
I have to admit that I'm not much wiser after studying them, but I at least have a couple of questions / comments:
- your shared `action_resume_bot` does a `UserUtteranceReverted`. However, in the log file I can't see a `rewind` event (aka `UserUtteranceReverted`) when this action is called (line 264 in the log file) - are you actually running the latest version of your actions?
- this somewhat relates to line 513 where the bot has to improvise because `There is no memorised next action`. Your ML policy (the alpaca really makes me curious 😆 ) then decides to do a `ACTION_LISTEN` event.

What I also did: I launched a Rasa instance with Redis LockStore + SQLite database and run a script which runs your described sequence repeatedly and checks if the bot responds. it worked every single time 🤯 

The script (sorry, it's a bit hacky):

```python
import uuid
from pprint import pprint

import requests

sender_id = "same"
base_url = "http://localhost:5005"


def test_prediction_after_resuming():
    global sender_id
    for _ in range(100):
        sender_id = uuid.uuid4().hex
        send_sequence()


def send_sequence():
    _send_message("hi")
    _print_tracker()
    _send_message("bye")
    _send_message("bye")
    _send_message("bye")
    _resume()
    # _print_tracker()
    assert _send_message("bye").json()[0]["text"] == 'Bye'


def _send_message(text):
    r = requests.post(
        f"{base_url}/webhooks/rest/webhook", json={"sender": sender_id, "message": text}
    )
    pprint(r.json())
    return r


def _print_tracker():
    r = requests.get(f"{base_url}/conversations/{sender_id}/tracker", params={"include_events": "APPLIED"})
    pprint(r.json())


def _resume():
    requests.post(
        f"{base_url}/conversations/{sender_id}/execute",
        json={"name": "action_resume_bot"},
    )
```

I ran this on a modified version of the project which is generated by `rasa init`. I changed the first two stories to:

```
## happy path
* greet
  - action_pause
* mood_great
  - utter_happy

## sad path 1
* greet
  - action_pause 
* mood_unhappy
  - utter_cheer_up
  - utter_did_that_help
* affirm
  - utter_happy
```

`action_pause` returns an `ConversationPause()` event (make sure to add the two new actions to your domain file in case you want to try this to)

**Comment by nbeuchat** [comment]
Just to answer regarding the `UserUtteranceReverted`: it will only be triggered if the user sent some messages while the bot was paused, I suppose @cristianmtr didn't do that in the example he sent you. 

Eheheh, cool policy name isn't it :D It's basically the `KerasPolicy` with an early stopping mechanism, nothing fancy! (and Alpaca is the name of the company eheheh!)

**Comment by wochinge** [comment]
> Just to answer regarding the UserUtteranceReverted: it will only be triggered if the user sent some messages while the bot was paused, I suppose @cristianmtr didn't do that in the example he sent you.

This is interesting! `UserUtteranceReverted` will revert all actions until the last user message, and then the message itself. if you omit the `UserUtteranceReverted` then your tracker will contain the `action_resume_bot` as event in the conversation history. This is unknown state for your story prediction and hence the bot does something wrong.

Options I see:
- revert the action when running `action_resume_bot`
- directly append the `ConversationResume()` event via https://rasa.com/docs/rasa/api/http-api/#operation/addConversationTrackerEvents
- use [this endpoint](https://rasa.com/docs/rasa/api/http-api/#operation/triggerConversationIntent) to trigger the `action_resume_bot` with an intent. This has the advantage that you can write stories for it so that the model learns how to deal with it

**Comment by nbeuchat** [comment]
The problem we've had and that we fixed with the `UserUtteranceReverted` is that we may have dozen of user messages without any action in between (as the bot was paused). We can't put these in the training data otherwise the bot will learn never to answer anything ;-) The idea was to revert every user intents until the resume (which happened to work with a single `UserUtteranceReverted`, it's probably not meant to!)

> use this endpoint to trigger the action_resume_bot with an intent. This has the advantage that you can write stories for it so that the model learns how to deal with it
I don't think it's possible, is it? The bot is still paused so not even the mapping policy will work. Or does it ignore the paused status? @cristianmtr opened a suggestion for that by the way #5630 
In any case, we cannot use it as we are using a hack to support multiple Facebook pages with a single bot but it does not allow us to use this new endpoint, unfortunately.

The thing we have trouble understanding is that it also fails if the message is supposed to trigger an action with Mapping Policy. So, even if our stories are wrong, it should still work for these specific intents no?

**Comment by wochinge** [comment]
> The thing we have trouble understanding is that it also fails if the message is supposed to trigger an action with Mapping Policy

@nbeuchat I assume that is because the `MappingPolicy` only jumps in, in case we had an `ACTION_LISTEN` event before it. 

> The idea was to revert every user intents until the resume (which happened to work with a single UserUtteranceReverted, it's probably not meant to!)

That shouldn't work. A `UserUtteranceReverted` only reverts 1 user message. so you should actually check how many messages where there since the `ConversationPause` event. Might also make sense to restart the conversation completeley.
